From a849f1e690ebab5a83988f51720ef15a77e34418 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:31 -0400
Subject: [usb] Revert: xhci: Set scatter-gather limit to avoid failed block writes

Message-id: <1410887463-20674-260-git-send-email-dzickus@redhat.com>
Patchwork-id: 95311
O-Subject: [RHEL7 PATCH 259/591] Revert "xhci: Set scatter-gather limit to avoid failed block writes."
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 1386ff75797a187df324062fb4e929152392da88
Author: Sarah Sharp <sarah.a.sharp@linux.intel.com>
Date:   Fri Jan 31 11:45:02 2014 -0800

    Revert "xhci: Set scatter-gather limit to avoid failed block writes."

    This reverts commit f2d9b991c549f159dc9ae81f77d8206c790cbfee.

    We are ripping out commit 35773dac5f862cb1c82ea151eba3e2f6de51ec3e "usb:
    xhci: Link TRB must not occur within a USB payload burst" because it's a
    hack that caused regressions in the usb-storage and userspace USB
    drivers that use usbfs and libusb.  This commit attempted to fix the
    issues with that patch.

    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
    Cc: stable@vger.kernel.org #3.12
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index e42e077..4101ab9 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -4730,8 +4730,8 @@ int xhci_gen_setup(struct usb_hcd *hcd, xhci_get_quirks_t get_quirks)
  struct device  *dev = hcd->self.controller;
  int   retval;
 
- /* Limit the block layer scatter-gather lists to half a segment. */
- hcd->self.sg_tablesize = TRBS_PER_SEGMENT / 2;
+ /* Accept arbitrarily long scatter-gather lists */
+ hcd->self.sg_tablesize = ~0;
 
  /* XHCI controllers don't stop the ep queue on short packets :| */
  hcd->self.no_stop_on_short = 1;
diff --git a/drivers/usb/host/xhci.h b/drivers/usb/host/xhci.h
index f841663..24344aa 100644
--- a/drivers/usb/host/xhci.h
+++ b/drivers/usb/host/xhci.h
@@ -1279,7 +1279,7 @@ union xhci_trb {
  * since the command ring is 64-byte aligned.
  * It must also be greater than 16.
  */
-#define TRBS_PER_SEGMENT 256
+#define TRBS_PER_SEGMENT 64
 /* Allow two commands + a link TRB, along with any reserved command TRBs */
 #define MAX_RSVD_CMD_TRBS (TRBS_PER_SEGMENT - 3)
 #define TRB_SEGMENT_SIZE (TRBS_PER_SEGMENT*16)
-- 
1.7.1